synapse/handlers/room_batch.py

```diff
@@ -213,9 +213,6 @@ async def persist_state_events_at_start(
                     room_id=room_id,
                     action=membership,
                     content=event_dict["content"],
-                    # Mark as an outlier to disconnect it from the normal DAG
-                    # and not show up between batches of history.
-                    outlier=True,
```

Comment:
Removing these as `outliers` because [`filter_events_for_client`](https://github.com/matrix-org/synapse/blob/e0bb2681340752f2716f1f386139ca37c53f26cd/synapse/visibility.py#L44) [filters out all `outlier` state](https://github.com/matrix-org/synapse/blob/e0bb2681340752f2716f1f386139ca37c53f26cd/synapse/visibility.py#L79-L81) except for out-of-band membership.

This state chain is still disconnected from the normal DAG because the first event has empty `prev_events` so they state just floats in a chain of its own.

![](https://user-images.githubusercontent.com/558581/160732053-47ec9ac5-b11d-4171-9c08-652cbeca5f25.png)

Reply:
I'm struggling to understand how the diagram relates to the change here. Which of the events on the diagram correspond to this code path?

Reply:
> I'm struggling to understand how the diagram relates to the change here. Which of the events on the diagram correspond to this code path?

Just giving it as a reference. It shows that the state chain floats by no `prev_events` and that the state chain is disconnected from the historical message chain.

Reply:
ok, so this code is handling the events in the yellow boxes labelled `batch0`, `batch1`, `batch2` ?

Reply:
`persist_state_events_at_start` is handling the state events **above the yellow boxes** (`m.room.member` events in the diagram). The historical state chain.

The yellow boxes are the historical messages chains

Reply:
> persist_state_events_at_start is handling the state events above the yellow boxes (m.room.member events in the diagram). The historical state chain.

well in that case, this seems wrong. those events seem to be random state events where we haven't wired them up to the event dag? why would we have the state at those events?

Either I'm severely understanding your diagram, or something is odd here.

Reply:
> well in that case, this seems wrong. those events seem to be random state events where we haven't wired them up to the event dag?

Correct, they float in their own chain.

> why would we have the state at those events?

But we technically do have the state for them because we're manually providing it via `state_event_ids`